### PR TITLE
Simplify catch with 'when'

### PR DESCRIPTION
**Pull request by jnyrup**
Spotted some places where exception handling could be simplified with `when`.

### CODE REVIEW

File: FluentFTP/Streams/FtpSocketStream.cs [hunk 1]
```diff
@@ -1097,18 +1087,14 @@ public class FtpSocketStream : Stream, IDisposable {
 					await EnableCancellation(m_socket.ConnectAsync(ipad, port), timeoutSrc.Token, () => DisposeSocket());
 				}
 			}
-			catch (SocketException ex) {
-				if (ex.SocketErrorCode == SocketError.OperationAborted ||
-					ex.SocketErrorCode == SocketError.TimedOut) {
-					throw new TimeoutException("Timed out trying to connect!");
-				}
-				throw;
+			catch (SocketException ex) when (ex.SocketErrorCode is SocketError.OperationAborted or SocketError.TimedOut) {
```
Comment (robinrodricks) [inline FluentFTP/Streams/FtpSocketStream.cs:1090]: would this new statement be perfectly equivalent in behaviour to the old one? I'm asking for the case at line 1105.

Comment (FanDjango) [reply FluentFTP/Streams/FtpSocketStream.cs:1090]: @jnyrup My intention in the original code was to rethrow to a different kind of exception in those special cases. You have eagle eyes, @robinrodricks.

Comment (jnyrup) [reply FluentFTP/Streams/FtpSocketStream.cs:1090]: Before:
* catch all `SocketException`
  * if `OperationAborted` or `TimedOut` throw `TimeoutException`
  * else rethrow

After:
* catch `SocketException` if `OperationAborted` or `TimedOut` and throw `TimeoutException`
* else go to next catch handler;

I think these should be equivalent?

Just for inspiration, to exemplify what you can do with pattern matching, you can use the syntax to combine both catches to a single.

```cs
catch (Exception ex) when (ex is SocketException { SocketErrorCode: SocketError.OperationAborted or SocketError.TimedOut } or ObjectDisposedException) {
	throw new TimeoutException("Timed out trying to connect!");
}
```

Comment (FanDjango) [reply FluentFTP/Streams/FtpSocketStream.cs:1090]: @jnyrup This is the one we are looking at, right?

```
			catch (SocketException ex) {
				if (ex.SocketErrorCode == SocketError.OperationAborted ||
					ex.SocketErrorCode == SocketError.TimedOut) {
					throw new TimeoutException("Timed out trying to connect!");
				}
				throw;
			}
```
and after applying the change we have:
```
			catch (SocketException ex) when (ex.SocketErrorCode is SocketError.OperationAborted or SocketError.TimedOut) {
}
```
We are losing the `throw new TimeoutException("Timed out trying to connect!");` IMO. Or not?

Comment (jnyrup) [reply FluentFTP/Streams/FtpSocketStream.cs:1090]: Have a look at the new file, I think it's the diff that might be confusing.

The new code is
https://github.com/robinrodricks/FluentFTP/blob/e2a35c38dea517124a727bd28651c38dfc2f8179/FluentFTP/Streams/FtpSocketStream.cs#L1090-L1092
and
https://github.com/robinrodricks/FluentFTP/blob/e2a35c38dea517124a727bd28651c38dfc2f8179/FluentFTP/Streams/FtpSocketStream.cs#L1106-L1108

Comment (FanDjango) [reply FluentFTP/Streams/FtpSocketStream.cs:1090]: You are right, 

>  I think it's the diff that might be confusing.

It confuses me totally. I would need to see the result of the change in its entirety to understand. And there is a complication by (even nested) conditional compiles and a stupid common catch clause for both cond branches.

I trust you and will try to understand it when I see the finished code.

Comment (FanDjango) [reply FluentFTP/Streams/FtpSocketStream.cs:1090]: Which I have now looked at. I really don't understand the diff, but the new code looks totally good.

Comment (FanDjango) [reply FluentFTP/Streams/FtpSocketStream.cs:1090]: I made the changes on a branch of the current master: (sorry, screenshot):

image

So I still don't understand the first half of the PR diff. But the result is AOK.

Comment (FanDjango) [reply FluentFTP/Streams/FtpSocketStream.cs:1090]: Sorry for the trouble I might have caused.

**Pull request merged**